src/main/java/swinglib/ErrorMessageBuilder.java

```diff
@@ -0,0 +1,117 @@
+package swinglib;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import java.awt.GraphicsEnvironment;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.swing.JOptionPane;
+
+import org.apache.commons.math3.util.Pair;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+
+
+/**
+ * Swing builder for an error message pop-up (Uses JOptionPane). Example usages:
+ * <code><pre>
+ *   JOptionPaneBuilder.builder()
+ *     .title("Something Crashed")
+ *     .message("Error Details")
+ *     .option("ok", () -> {})
+ *     .option("quit", () -> System.exit(0))
+ *     .buildAndShow();
+ * </pre></code>
+ * Default error dialog with an 'ok' button:
+ * <code><pre>
+ *   JOptionPaneBuilder.builder()
+ *     .message("error details")
+ *     .buildAndShow();
+ * </pre></code>
+ */
+public class ErrorMessageBuilder {
+
+  private String title = "Error";
+  private String message;
+  private List<Pair<String, Runnable>> actions = new ArrayList<>();
+
+  private ErrorMessageBuilder() {
+  }
+
+  public static ErrorMessageBuilder builder() {
+    return new ErrorMessageBuilder();
+  }
+
+  public ErrorMessageBuilder message(String message) {
+    this.message = message;
+    return this;
+  }
+
+  /**
+   * Constructs a Swing JDialog using current builder values.
+   * Values that must be set: title, contents
+   */
+  public void buildAndShow() {
+    checkState(!GraphicsEnvironment.isHeadless(), "JOptionPane throws an exception in headless environments");
+    checkNotNull(message);
+
+    if (actions.isEmpty()) {
+      actions.add(new Pair<>("OK", () -> {
+      }));
+    }
+
+    // First option element to JOptionPane is the right most button, right to left ordering.
+    // It is more natural for left-to-right ordering of options to be specified, so we
+    // will reverse the option ordering to allow clients to specify left to right ordering.
+
+    List<Pair<String, Runnable>> swingOptionsRightToLeft = reverse(actions);
```

Comment:
It's an API thing, that would mean all clients would need to know about that detail. Most swing components are added left to right, so clients knowing that this *one* API is special and adds right to left is confusing/error prone.

The swing builder is a good place to hide this detail, this way we do not expose such inconsistencies to the rest of the codebase, and we can keep that codebase cleaner.

For consideration, this example:
```
.option("ok", () -> ...)
.option("cancel", () -> ...)
```

For me that most naturally has "ok" then "cancel", with "OK" on the left and "OK" as the default. I thought it confusing for this code to list "cancel" then "ok" and have the last option specified be the default instead of the first.


*one note* we are not actually sorting the options, we are just reversing the ordering. The ordering is specified by client code using the `ErrorMessageBuilder.builder()`

Reply:
What I actually meant was why isn't `actions.add(0, obj)` being called instead of `actions.add(obj)`?